quai/downloader/fetcher.go

```diff
@@ -0,0 +1,233 @@
+package downloader
+
+import (
+	"math/big"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/dominant-strategies/go-quai/common"
+	"github.com/dominant-strategies/go-quai/core/types"
+	"github.com/dominant-strategies/go-quai/ethdb"
+	"github.com/dominant-strategies/go-quai/log"
+	"github.com/dominant-strategies/go-quai/trie"
+	"github.com/pkg/errors"
+)
+
+const (
+	// c_fetchTimeout is the timeout for fetching a structure from the network
+	c_fetchTimeout = 30 * time.Second
+
+	// c_numTrieWorkers is the number of concurrent workers to fetch trie nodes
+	c_numTrieWorkers = 10
+
+	// c_fetchRetries is the number of times to retry fetching a trie node
+	c_fetchRetries = 3
+
+	// c_fetcherQueueSize is the size of the queue for trie node hashes
+	c_fetcherQueueSize = 1000000
+)
+
+type fetcher struct {
+	p2pNode     P2PNode
+	queue       chan common.Hash     // queue for trie node hashes that need to be fetched
+	fetched     map[common.Hash]bool // map to keep track of fetched or scheduled hashes to avoid duplicates
+	mu          sync.Mutex           // mutex to protect the fetched map
+	db          ethdb.Database       // local database to save the fetched trie nodes
+	quitCh      chan struct{}        // channel to signal the fetcher to stop
+	pendingJobs int64                // Atomic counter to track pending jobs
+	doneCh      chan struct{}        // channel to signal workers that all jobs are done
+}
+
+// FetchBlock fetches a single block by its number.
+func (d *fetcher) fetchBlock(loc common.Location, blockNumber *big.Int) (*types.Block, error) {
```

Comment:
again, I am surprised by two things:
1) why do we need this at all? afaik block fetching was already working
2) if we do need this, why is it in this commit. It is unrelated to snap sync messaging